File: src/agents/engineeringExecution/gitOps.ts
import { runOrThrow } from './shell.js';

export interface AssetFile {
  /** Absolute (or worktree-cwd-relative) path to the file on disk. */
  path: string;
  /** Flat filename the asset is stored/served under — no directories. */
  name: string;
}

/** No method here ever touches `main`/the default branch, force-pushes, or
 * merges anything. Branch/commit/push only, on the caller-supplied isolated
 * branch name, inside the caller-supplied worktree. */
export interface GitOps {
  createBranch(worktreePath: string, branchName: string): Promise<void>;
  /** Returns false if there was nothing to commit. */
  commitAll(worktreePath: string, message: string): Promise<boolean>;
  push(worktreePath: string, branchName: string): Promise<void>;
  diffStat(worktreePath: string, baseRef: string): Promise<string>;
  /**
   * Publishes files as a standalone orphan branch — via git plumbing
   * directly against the object database, so it never touches the working
   * tree or index of whatever branch is currently checked out in
   * `worktreePath`. Used to host PR screenshots/GIFs so they're linkable
   * (via raw.githubusercontent.com) without ever being part of the code
   * PR's own diff or commit history.
   */
  publishAssetBranch(worktreePath: string, branchName: string, files: AssetFile[]): Promise<void>;
}

export class GitCliOps implements GitOps {
  async createBranch(worktreePath: string, branchName: string): Promise<void> {
    await runOrThrow(['git', 'checkout', '-b', branchName], { cwd: worktreePath });
  }

  async commitAll(worktreePath: string, message: string): Promise<boolean> {
    await runOrThrow(['git', 'add', '-A'], { cwd: worktreePath });
    const status = await runOrThrow(['git', 'status', '--porcelain'], { cwd: worktreePath });
    if (!status.output.trim()) return false;
    await runOrThrow(['git', 'commit', '-m', message], { cwd: worktreePath });
    return true;
  }

  async push(worktreePath: string, branchName: string): Promise<void> {
    await runOrThrow(['git', 'push', '-u', 'origin', branchName], { cwd: worktreePath, timeoutMs: 180_000 });
  }

  async diffStat(worktreePath: string, baseRef: string): Promise<string> {
    const result = await runOrThrow(['git', 'diff', '--stat', baseRef], { cwd: worktreePath });
    return result.output;
  }

  async publishAssetBranch(worktreePath: string, branchName: string, files: AssetFile[]): Promise<void> {
    if (files.length === 0) return;
    const entries: string[] = [];
    for (const file of files) {
      const hashed = await runOrThrow(['git', 'hash-object', '-w', file.path], { cwd: worktreePath });
      entries.push(`100644 blob ${hashed.output.trim()}\t${file.name}`);
    }
    const tree = await runOrThrow(['git', 'mktree'], { cwd: worktreePath, input: entries.join('\n') + '\n' });
    const commit = await runOrThrow(
      ['git', 'commit-tree', tree.output.trim(), '-m', `Screenshots for ${branchName}`],
      { cwd: worktreePath },
    );
    await runOrThrow(['git', 'push', 'origin', `${commit.output.trim()}:refs/heads/${branchName}`], {
      cwd: worktreePath,
      timeoutMs: 180_000,
    });
  }
}

export class MockGitOps implements GitOps {
  calls: Array<[string, ...unknown[]]> = [];
  hasChanges = true;

  constructor(private fakeDiffStat = ' src/example.ts | 4 ++--\n 1 file changed, 2 insertions(+), 2 deletions(-)') {}

  async createBranch(worktreePath: string, branchName: string): Promise<void> {
    this.calls.push(['createBranch', worktreePath, branchName]);
  }

  async commitAll(worktreePath: string, message: string): Promise<boolean> {
    this.calls.push(['commitAll', worktreePath, message]);
    return this.hasChanges;
  }

  async push(worktreePath: string, branchName: string): Promise<void> {
    this.calls.push(['push', worktreePath, branchName]);
  }

  async diffStat(worktreePath: string, baseRef: string): Promise<string> {
    this.calls.push(['diffStat', worktreePath, baseRef]);
    return this.fakeDiffStat;
  }

  async publishAssetBranch(worktreePath: string, branchName: string, files: AssetFile[]): Promise<void> {
    this.calls.push(['publishAssetBranch', worktreePath, branchName, files]);
  }
}

File: legacy-python/work_agent/execution/git_ops.py
from __future__ import annotations

from typing import Protocol

from . import shell


class GitOps(Protocol):
    """No method here ever touches `main`/the default branch, force-pushes,
    or merges anything. Branch/commit/push only, on the caller-supplied
    isolated branch name, inside the caller-supplied worktree."""

    def create_branch(self, worktree_path: str, branch_name: str) -> None: ...

    def commit_all(self, worktree_path: str, message: str) -> bool:
        """Returns False if there was nothing to commit."""
        ...

    def push(self, worktree_path: str, branch_name: str) -> None: ...

    def diff_stat(self, worktree_path: str, base_ref: str = "HEAD") -> str: ...


class GitCliOps:
    """Real git operations via subprocess. Safe by construction: every call
    is scoped to `worktree_path`, which the caller is expected to have
    created via WorktreeManager (never the primary checkout)."""

    def create_branch(self, worktree_path: str, branch_name: str) -> None:
        shell.run_or_raise(["git", "checkout", "-b", branch_name], cwd=worktree_path)

    def commit_all(self, worktree_path: str, message: str) -> bool:
        shell.run_or_raise(["git", "add", "-A"], cwd=worktree_path)
        status = shell.run_or_raise(["git", "status", "--porcelain"], cwd=worktree_path)
        if not status.output.strip():
            return False
        shell.run_or_raise(["git", "commit", "-m", message], cwd=worktree_path)
        return True

    def push(self, worktree_path: str, branch_name: str) -> None:
        shell.run_or_raise(["git", "push", "-u", "origin", branch_name], cwd=worktree_path, timeout=180)

    def diff_stat(self, worktree_path: str, base_ref: str = "HEAD") -> str:
        result = shell.run_or_raise(["git", "diff", "--stat", base_ref], cwd=worktree_path)
        return result.output


class MockGitOps:
    """Records calls, never touches disk or a remote. For orchestration tests."""

    def __init__(self, fake_diff_stat: str = " src/example.ts | 4 ++--\n 1 file changed, 2 insertions(+), 2 deletions(-)"):
        self.calls: list[tuple] = []
        self._fake_diff_stat = fake_diff_stat
        self.has_changes = True

    def create_branch(self, worktree_path: str, branch_name: str) -> None:
        self.calls.append(("create_branch", worktree_path, branch_name))

    def commit_all(self, worktree_path: str, message: str) -> bool:
        self.calls.append(("commit_all", worktree_path, message))
        return self.has_changes

    def push(self, worktree_path: str, branch_name: str) -> None:
        self.calls.append(("push", worktree_path, branch_name))

    def diff_stat(self, worktree_path: str, base_ref: str = "HEAD") -> str:
        self.calls.append(("diff_stat", worktree_path, base_ref))
        return self._fake_diff_stat
